### DISCUSSION TITLE
Testing knowledge augmentation of LLMs - RAG vs. Context Stuffing

### DISCUSSION THREAD

**Discussion by anupamck** [BeyondQuality/beyondquality#4] [Research]
One of the most promising use-cases for LLMs is to connect them to a knowledge base so that they can be used as a conversational interface to accessing that knowledge. E.g., linking an LLM to a company's documentation portal to answer and retrieve the right documents needed to answer specific questions. 

There are several ways to augment LLM knowledge. In this research tasks, I am specifically exploring two of these - Retrieval Augmented Generation (RAG) and context stuffing. 

## What I have already done

Context stuffing involves connecting an LLM directly with the knowledge base (in effect, including the entire knowledge base within the LLM's prompt). While this is the quickest to set up, LLMs have limited context windows, which makes them unreliable when stuffed with large knowledge bases. I did some exploratory testing of a context stuffed model and documented the results [here](https://anukrit.de/articles/context-stuffing/). 

## What I want to do next

In the next part of my research, I want to go deeper. I am in the process of [implementing a RAG solution](https://github.com/anupamck/rag101). This is still a work in progress. I wish to extend this by feeding the model posts from my [old blog](https://anupamobserved.wordpress.com/), and to design an experiment to compare context stuffing with a RAG implementation.  

## Inputs & collaboration are welcome

Let me know if you already know something about this topic, or would like to collaborate on this piece of research. 

## Sources

- [Excellent paper](https://writings.stephenwolfram.com/2023/02/what-is-chatgpt-doing-and-why-does-it-work/) by Stefan Wolfram on how LLMs work
- [A somewhat detailed introduction to RAG](https://cameronrwolfe.substack.com/p/a-practitioners-guide-to-retrieval?utm_source=profile&utm_medium=reader2)   
- [A tutorial from Langchain](https://python.langchain.com/docs/tutorials/rag/), which I have followed to build a working POC. 
- [This excellent Youtube video playlist](https://www.youtube.com/playlist?list=PLfaIDFEXuae2LXbO1_PKyVJiQ23ZztA0x) which explains the Langchain tutorial in more detail. 

## Credits

To [Rahul Verma](https://www.linkedin.com/in/rahul-verma-india/) for inspiring this approach and giving me some valuable guidance.

**Comment by sharovatov** [comment]
The idea of a conversational agents to some body of knowledge is **extremely** interesting.

Papers I've read so far on the topic:
- [Retrieval-Augmented Generation for Knowledge-Intensive NLP Tasks](https://arxiv.org/abs/2005.11401). Summary: [on QA 'language-generating' tasks] RAG improves factual accuracy compared to relying only on pre-trained LLM weights + it allows external KBs to be updated independently of retraining the LLM (which might be prohibitively expensive)
- [Understanding Knowledge Drift in LLMs through Misinformation](https://arxiv.org/abs/2409.07085). Summary: in the concrete dialogue with the LLM, the more incorrect information I submit, the higher the chat the model will "drift" (i.e. reiterate on the provided misinformation), so the chat has to be recreated from scratch. 
- [continual learning and catastrophic forgetting](https://arxiv.org/pdf/2403.05175v1) Sumary: continuous learning for an LLM isn't possible yet, retraining **or RAG** is needed.

Challenges:
- **Hallucinations**. As far as I understand, RAG significantly reduces but doenst eliminate hallucinations.
- **Reasoning drift**. Again, afaiu, RAG helps but doesn't solve this.

**I'd love to participate in this!**

**Reply by anupamck** [reply]
Thanks @sharovatov! The papers you have linked outline the advantages and limitations of RAG. In the process, I wish to get a better understanding of:
- LLMs in general
- RAG models in particular
- The capabilities and the limitations of RAG (through testing)
- The tooling and workflow needed to deploy a RAG solution

 I'd love to brainstorm and see how we can take this research task forward.  I'll reach out to you.

**Reply by anupamck** [reply]
During the last week, I have discovered that when we attach long documents to LLMs like ChatGPT, they implement [RAG under the hood](https://help.openai.com/en/articles/8843948-knowledge-in-gpts). I had earlier assumed that these LLMs were context stuffed. 

My motivation for comparing RAG with context stuffing was to justify the effort required in building a custom RAG pipeline. Seeing as how this effort is trivial (offered out of the box by LLMs providers), I decided to pivot my research to RAG evaluation. 

I am hereby closing this discussion, and I will start a new one on RAG assessments.
